Embedded Arimo font widths and used that when fonts are missing

huzjakd opened the pull request:
This fixes this bug: https://github.com/mozilla/pdf.js/issues/11162
And answers my question: https://github.com/mozilla/pdf.js/issues/11175

I embedded Arimo font widths and used them to fall back to a default font when the fonts are missing.

It passes lint tests but doesn't pass 8-10 others.
When comparing screenshots of failed tests the only differences are a few pixels that are not visible by the naked eye.
Also in the attached file, you don't normally see the number 3 in "Page 13" but with this fix, you see it which causes the test to fail.
[pr4922.pdf](https://github.com/mozilla/pdf.js/files/3681426/pr4922.pdf)

huzjakd committed:
Embedded Arimo font widths and used that when fonts are missing

timvandermeij commented:
I can't really review this implementation because I'm not too familiar with the font conversion code so I'll leave that for others, but I think in general the idea of falling back to a default font is better than not having any font at all.

Snuffleupagus commented on src/core/evaluator.js:
```diff
@@ -723,29 +723,70 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
-        warn('fontRef not available');
-        return errorFont();
+        // warn('fontRef not available');
+        // return errorFont();
```
Please note that commented-out code isn't allowed in general, and in this case you very much want to print the warning since there's simply no guarantee that things will actually work particularly well in general with this sort work-around.

Also, an sending an `UnsupportedFeature` notification would seem appropriate as well.

Snuffleupagus commented on src/core/evaluator.js:
```diff
@@ -723,29 +723,70 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
+        // warn('fontRef not available');
+        // return errorFont();
+
+        // fallback to some default font
```
Nit: Please ensure that added comments always are complete sentences, i.e. starting with an upper-case letter and end with a dot.

Snuffleupagus commented on src/core/evaluator.js:
```diff
@@ -723,29 +723,70 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
+        // return errorFont();
+
+        // fallback to some default font
+        let widthsWithNulls = [null, null, null, null, null, null, null,
```
For the general PDF.js library, you should use widths corresponding a *standard font* and we also want to avoid defining this data inline.

Please find already embedded *standard font* widths in https://github.com/mozilla/pdf.js/blob/master/src/core/metrics.js, where the `Helvetica` ones seem like a good choice for the fallback font being constructed here.

Snuffleupagus commented on src/core/evaluator.js:
```diff
@@ -723,29 +723,70 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
+        fallbackFontProperties = {
+          fallbackFont: true,
+          firstChar: 0,
+          name: 'BCDEEE+Arimo',
```
This name seems almost like it's taken from an *actual* PDF file, which doesn't seem entirely appropriate for a *general* fallback font; how about simply 'PDFJS-FallbackFont' or similar instead.

Snuffleupagus commented on src/core/evaluator.js:
```diff
@@ -723,29 +723,70 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
+            widthsWithoutNulls[i] = widthsWithNulls[i];
+          }
+        }
+        fallbackFontProperties = {
```
Please note that a plain `Object` isn't an appropriate data-structure here, since it for one seem to cause *a lot* of unwanted special-casing throughout the font parsing code below.

Instead you need to create a *valid* font dictionary here, with dummy entries as needed, to ensure that the fallback font dictionary can be parsed as-is by the relevant code below.

Snuffleupagus reviewed (changes requested):
I've tried to do an initial review pass, please see the inline comments.

Please note that the single most important point here is that, for reasons of both readability and maintainable of the code, you need to create a *valid* font dictionary such that all of the special cases (in the various font parsing functions) can be removed.

huzjakd commented:
The way I see it when some font is not embedded in the pdf, the pdf.js always logs a warning and returns the ErrorFont and doesn't display anything in the pdf. At this point, I just want to embed/construct my own font and use it to display something anyway.

huzjakd committed:
Merge branch 'master' into bugFix-missingFonts

huzjakd committed:
Different approach and changes acording to the reviewer

I took a different approach here by generating a dictionary and using it for a fallback font. So most of the old "special cases" code was removed and left with only a few lines of code to achieve the same result.

timvandermeij commented:
Closing since the pull request above built upon the changes in this pull request and was merged. Thank you for contributing this!
